### PR TITLE
fix: drain tunnel gateway routes and sessions on shutdown

### PR DESCRIPTION
**Pull request by ThomasRooney**
## Summary

Introduces a single-writer `routeReconciler`: one per-pod goroutine owns every mutating
route-store call (publish, connection snapshots, unpublish, global revoke). Handlers mark a
tunnel dirty and nudge; the reconciler reads live registry state at write time, batch-refreshes
all live tunnels on one ticker (replacing the per-session refresher goroutines), gates
publishes on `IsActive`, and prunes tracking for cleanly-removed tunnels.

Because all writes are serialized through one goroutine, graceful shutdown becomes the
reconciler's final act: on SIGTERM the registry stops admitting agent connects (checked again
under the lock at registration), the reconciler performs one bounded batched cleanup of every
route this pod may own, the forward server drains in-flight requests, agent sessions close
concurrently, and `main` blocks until the sequence completes — all within the existing 25s
budget. Publish-after-cleanup races are impossible by construction rather than coordinated
away.

Tests are black-box at the gateway boundary (real WebSocket agent + yamux sessions against an
in-memory store): post-drain store emptiness with no writes after drain returns, 503 on
connect during drain, in-flight forwards completing through shutdown, TTL refresh cadence,
revocation across multiple owners, and a concurrent-churn convergence property under the race
detector.

## Motivation

The gateway exited without unpublishing its routes, so every rollout left pod-IP routes in
Redis for up to the 30s TTL while the pod was already gone. gram-server dialed the dead IP and
each affected tunneled request hung ~30s before surfacing a 502, producing an error burst on
every deploy. Draining removes the stale-route window; agents re-home to surviving pods in
~0.5–2s using their existing reconnect backoff. Complements #5888 (fast dial timeout) and
#5930 (dead-route eviction on dial failure), which cover unplanned pod death.

closes AIM-166

### CODE REVIEW

File: tunnel/gateway/gateway.go [hunk 1]
```diff
@@ -102,7 +110,175 @@ func (g *Gateway) ActiveSessions() int { return g.reg.activeSessions() }
...
+				slog.Int("tunnels", len(snapshot.tunnelIDs)), slog.Int("sessions", snapshot.sessionCount))
+		} else {
+			g.logger.WarnContext(ctx, "tunnel gateway drain publisher wait expired",
+				slog.Int("tunnels", len(snapshot.tunnelIDs)),
+				slog.Int("sessions", snapshot.sessionCount),
+				slog.Any("error", publisherWaitErr))
+		}
+
+		cleanupCompleted := g.cleanupOwnedTunnels(ctx, snapshot.tunnelIDs, "initial")
+		if !publishersStopped && cleanupCompleted {
```
Comment (bflad) [inline tunnel/gateway/gateway.go]: **High**: the final sweep only runs when the publisher join times out, but `beginDrain` cancels in-flight publishes, and go-redis can return `ctx.Err()` after the `ZAdd` bytes were already flushed to the socket. In that interleaving `waitForPublishers` succeeds within the 2s budget, only the initial sweep runs, and the delayed `ZAdd` lands after the sweep's `ZRem`, resurrecting the dead pod's route for up to the 30s TTL. The `blockedPublishStore` test only covers the join-timeout path, not this early-return variant.

Quick fix: run the `"final"` `cleanupOwnedTunnels` sweep unconditionally (it is idempotent and cheap when Redis is healthy) instead of gating it on `!publishersStopped`.

File: tunnel/cmd/tunnel-gateway/main.go [hunk 2]
```diff
@@ -78,12 +79,22 @@ func main() {
...
 
+	var shutdownOnce sync.Once
+	shutdownDone := make(chan struct{})
+	shutdown := func() {
+		shutdownOnce.Do(func() {
+			defer close(shutdownDone)
+			shutCtx, cancel := context.WithTimeout(context.Background(), 25*time.Second)
+			defer cancel()
+			gw.Drain(shutCtx)
+			_ = forwardSrv.Shutdown(shutCtx)
```
Comment (bflad) [inline tunnel/cmd/tunnel-gateway/main.go]: **High**: all four phases share the one 25s `shutCtx` and `forwardSrv.Shutdown` blocks until every in-flight forward finishes. A single long-lived SSE forward (the proxy sets `FlushInterval: -1` and there is no per-request timeout) eats the remaining ~24s, so `CloseSessions` then runs with an already-expired context and session teardown becomes fire-and-forget at process exit, racing SIGKILL.

Quick fix: give `forwardSrv.Shutdown` its own sub-deadline (e.g. `context.WithTimeout(shutCtx, 15*time.Second)`) so `CloseSessions` and `publicSrv.Shutdown` always keep a reserved slice of the budget.

File: tunnel/cmd/tunnel-gateway/main.go [hunk 3]
```diff
@@ -78,12 +79,22 @@ func main() {
...
+			gw.CloseSessions(shutCtx)
+			_ = publicSrv.Shutdown(shutCtx)
```
Comment (bflad) [inline tunnel/cmd/tunnel-gateway/main.go:98]: **High**: the public server used to shut down at signal time and now shuts down last, so `/healthz` keeps returning its unconditional 200 for the whole ~25s drain window while `handleConnect` 503s every agent. Anything keyed on that health check keeps steering re-homing agents at the dying pod, and each rejected connect burns a full jittered backoff cycle, stretching the advertised sub-2s re-home.

Quick fix: make `/healthz` drain-aware (return 503 once `isDraining()`), which lets the LB pull the pod immediately while the listener stays up. Keeping the listener up is fine for live sessions since `http.Server.Shutdown` never touches hijacked websocket connections.

File: tunnel/gateway/gateway.go [hunk 4]
```diff
@@ -223,50 +417,79 @@ func (g *Gateway) handleConnect(w http.ResponseWriter, r *http.Request) {
...
-		g.publishConnectionSnapshot(stateCtx, tunnelID, time.Now().UTC())
+		cancelPublish()
+		snapshotCtx, cancelSnapshot := routeOperationContext(context.Background())
+		g.publishConnectionSnapshot(snapshotCtx, tunnelID, time.Now().UTC())
+		cancelSnapshot()
+		return
+	}
+	// Retain ownership after either failure so shutdown drain retries both
+	// idempotent cleanup operations instead of forgetting possibly stale state.
+	if unpublishErr == nil && snapshotDeleteErr == nil {
```
Comment (bflad) [inline tunnel/gateway/gateway.go]: **Medium**: retaining ownership on failure is the right call, but nothing retries it until process shutdown. If Redis blips during a mass disconnect, every affected tunnel skips `finishCleanup`, and since agents normally re-home to another pod, no later path on this pod deletes the entry. `ownedTunnels` then grows for the process lifetime and a much later `Drain` spends its bounded 25s sweeping thousands of long-dead tunnels, which can starve cleanup of live routes.

Quick fix: add a bounded retry, e.g. a periodic janitor that re-attempts `Unpublish`/`DeleteConnectionOwner` for entries stuck in the retained state (or piggyback on the existing refresh ticker cadence).

File: tunnel/gateway/gateway.go [hunk 5]
```diff
@@ -223,50 +417,79 @@ func (g *Gateway) handleConnect(w http.ResponseWriter, r *http.Request) {
 // its own Publish after our Unpublish, so every interleaving converges on a
 // published route while live sessions exist.
 func (g *Gateway) cleanupSessionState(tunnelID string) {
```
Comment (bflad) [inline tunnel/gateway/gateway.go]: **Medium**: this function lost its single shared 5s budget, it is now up to four sequential 5s-bounded ops, so under Redis degradation each dying-session goroutine can block ~20s instead of 5s, quadrupling goroutine and connection lifetime exactly when Redis is already struggling. It is also not drain-aware, so during `CloseSessions` every woken connect handler re-issues `Unpublish`/`DeleteConnectionOwner` for tunnels the drain sweep just force-cleaned.

Quick fix: create one `routeOperationContext` at the top and pass it to all four ops to restore the shared budget, and early-return when `g.reg.isDraining()` since `cleanupOwnedTunnels` already owns that state.

File: tunnel/gateway/gateway.go [hunk 6]
```diff
@@ -170,6 +346,12 @@ func (g *Gateway) handleConnect(w http.ResponseWriter, r *http.Request) {
 	// or advance last_seen_at for a tunnel that never connected. On failure,
 	// close the session — the agent retries and re-activates.
 	if recorder, ok := g.keys.(ConnectionRecorder); ok {
+		if g.reg.isDraining() {
```
Comment (bflad) [inline tunnel/gateway/gateway.go]: **Medium**: this re-check narrows the window but is still TOCTOU, drain can begin between `isDraining()` returning false and the durable `MarkConnected` write, so a draining pod can still persist a fresh activation for a session that `reg.add` immediately rejects. It self-heals on re-home, but `TestConnectRechecksDrainBeforeMarkConnected` asserts `markConnectedCalls == 0` as if this were a guaranteed invariant when it only holds for one interleaving.

Quick fix: either document the check (and the test) as best-effort narrowing rather than a guarantee, or close the window properly by registering the pending mark with the registry so `beginDrain` waits for it, similar to the publisher tracking.

**Review by bflad — APPROVED** [review]
Approving, the drain sequencing and the test coverage here are solid and this should eliminate the deploy 502 bursts in the common case. I left inline comments on findings worth discussing: three edge cases that can undermine the drain (a late publish landing after the sweep, one 25s budget shared by all shutdown phases, and `/healthz` staying green while draining) and three smaller ones (retained ownership never retried, `cleanupSessionState` budget growth plus drain redundancy, and the `MarkConnected` drain re-check being best-effort). None of these block merging from my side and they could all be followups.

File: tunnel/gateway/gateway.go (hunk 1 above)
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go]: Superseded by the rewrite in 2f176a47e5: route publishing moved to a single-writer reconciler goroutine, and the drain sweep is now its unconditional final act — there is no conditional final-sweep path (and no second writer whose ZAdd could land after it).

File: tunnel/cmd/tunnel-gateway/main.go (hunk 2 above)
Comment (ThomasRooney) [reply tunnel/cmd/tunnel-gateway/main.go]: Fixed in 2f176a47e5: forwardSrv.Shutdown now runs under its own 15s sub-deadline so CloseSessions and publicSrv.Shutdown always keep a reserved slice of the 25s budget.

File: tunnel/cmd/tunnel-gateway/main.go (hunk 3 above)
Comment (ThomasRooney) [reply tunnel/cmd/tunnel-gateway/main.go:98]: Fixed in 2f176a47e5: /healthz returns 503 once draining (liveness shares the endpoint but needs 3 failures x 20s, well past the 25s drain, so no mid-drain kill; readiness flipping immediately is the point).

File: tunnel/gateway/gateway.go (hunk 4 above)
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go]: Fixed in 2f176a47e5: the rewrite tracks owned tunnels in the reconciler; each refresh tick now retries cleanup for non-live retained entries through the bounded worker pool and prunes on success, so failed cleanups converge at tick cadence instead of accumulating until shutdown.

File: tunnel/gateway/gateway.go (hunk 5 above)
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go]: The rewrite (2f176a47e5) deleted cleanupSessionState — disconnects just nudge the reconciler, whose cleanup now uses one shared 5s operation context for both ops, and post-drain nudges are dropped so nothing races the drain sweep.

File: tunnel/gateway/gateway.go (hunk 6 above)
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go]: Took option (a) in 2f176a47e5: the pre-MarkConnected check is documented as best-effort narrowing (reg.add stays authoritative; a lost race costs one spurious activation that self-heals on re-home), and the test pins only the deterministic drain-before-connect case.

File: tunnel/gateway/registry.go [hunk 7]
```diff
@@ -76,6 +111,135 @@ func (r *registry) add(tunnelID, sessionID string, s *yamux.Session, proxy http.
...
+		}
+	}
+	sort.Strings(snapshot.tunnelIDs)
+	return snapshot
+}
+
+func (r *registry) beginPublish(parent context.Context, tunnelID string) (context.Context, func(), bool) {
+	r.mu.RLock()
+	ownership := r.ownedTunnels[tunnelID]
+	if r.draining || ownership == nil {
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/registry.go]: Superseded by the single-writer rewrite (2f176a47e5): beginPublish/finishCleanup no longer exist — all store writes serialize through one reconciler goroutine, so no publish can be admitted after cleanup.

File: tunnel/gateway/gateway.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go]: Superseded by the rewrite (2f176a47e5): the drain sweep is the reconciler's unconditional final act over every tunnel it ever owned, regardless of earlier failures.

File: tunnel/gateway/registry.go [hunk 8]
```diff
@@ -19,9 +22,15 @@ const consumerSessionTTL = 5 * time.Minute
 // Multiple agents may share a tunnel key. Stable consumer keys stick to one live session;
 // requests without a consumer key round-robin across live sessions.
 type registry struct {
-	mu       sync.RWMutex
-	sessions map[string][]*sessEntry
-	rr       map[string]uint64 // round-robin cursor per tunnel
+	mu                sync.RWMutex
+	sessions          map[string][]*sessEntry
+	rr                map[string]uint64 // round-robin cursor per tunnel
+	ownedTunnels      map[string]*tunnelOwnership
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/registry.go]: Superseded by the rewrite (2f176a47e5): revocation is a synchronous reconciler command doing the global Delete/DeleteConnections, and owned-tunnel tracking is pruned on successful cleanup.

File: tunnel/gateway/gateway.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go]: Fixed in 2f176a47e5: refresh ticks retry cleanup for retained non-live entries via the bounded pool and prune on success — transient Redis failures converge at tick cadence.

File: tunnel/cmd/tunnel-gateway/main.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/cmd/tunnel-gateway/main.go]: Fixed in c943cd73d9: explicit phase budgets inside the 25s total — Drain, forward shutdown, and session close each get reserved sub-deadlines, so CloseSessions always runs with a live context.

File: tunnel/gateway/registry.go [hunk 9]
```diff
@@ -76,10 +85,45 @@ func (r *registry) add(tunnelID, sessionID string, s *yamux.Session, proxy http.
...
 
-func (r *registry) tunnelSessionCount(tunnelID string) int {
+func (r *registry) keyHash(tunnelID string) (string, bool) {
 	r.mu.RLock()
 	defer r.mu.RUnlock()
-	return len(r.sessions[tunnelID])
+
+	for _, entry := range r.sessions[tunnelID] {
+		if !entry.session.IsClosed() {
+			return entry.keyHash, true
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/registry.go]: Fixed in c943cd73d9: IsActive validation is per-session again (each session's own presented key hash); only inactive-hash sessions are closed and the route stays published while any active session remains.

File: tunnel/gateway/registry.go [hunk 10]
```diff
@@ -40,19 +42,26 @@ func newRegistry() *registry {
...
 		id:               sessionID,
+		keyHash:          keyHash,
 		session:          s,
 		proxy:            proxy,
 		connection:       connection,
 		activeSubstreams: 0,
 		consumerSessions: make(map[string]time.Time),
 	}
 	r.mu.Lock()
+	if r.draining {
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/registry.go:65]: Acknowledged as accepted best-effort per earlier reviewer discussion (see thread on the same check): reg.add remains the authoritative gate, a lost race costs one spurious durable activation that self-heals on re-home, and the code/test are documented accordingly. Serializing the marker write with admission isn't worth the coupling in this PR.

File: tunnel/e2e/tunnel_test.go [hunk 11]
```diff
@@ -354,6 +414,16 @@ func requireEventually(t *testing.T, d time.Duration, cond func() bool) {
...
 }
 
+func cleanupGateway(t *testing.T, gw *gateway.Gateway) {
+	t.Helper()
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		defer cancel()
+		gw.Drain(ctx)
+		gw.CloseSessions(ctx)
+	})
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/e2e/tunnel_test.go:431]: Fixed in c943cd73d9: the e2e cleanup helper gives CloseSessions its own fresh context.

File: tunnel/gateway/reconciler.go [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+		}
+
+		select {
+		case req := <-r.drain:
+			r.finishDrain(req, everOwned)
+			return
+		case <-r.nudges:
+			r.reconcile(r.lifetimeCtx, r.takeDirty(), false, everOwned)
+		case req := <-r.revoke:
+			if r.deleteTunnel(req.ctx, req.tunnelID) {
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/reconciler.go:185]: Fixed in c943cd73d9: failed global revoke deletes are retained as pending and both deletes retry on refresh ticks until they succeed.

File: tunnel/gateway/reconciler.go [hunk 13]
```diff
@@ -0,0 +1,352 @@
...
+			return
+		default:
+		}
+
+		select {
+		case req := <-r.drain:
+			r.finishDrain(req, everOwned)
+			return
+		case <-r.nudges:
+			r.reconcile(r.lifetimeCtx, r.takeDirty(), false, everOwned)
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/reconciler.go]: Fixed in c943cd73d9: publishes are suppressed for tunnels with a pending/completed revoke; suppression lifts only when a session with a verified active key hash appears.

File: tunnel/e2e/tunnel_test.go [hunk 14]
```diff
@@ -305,6 +309,62 @@ func TestTunnelRevoke(t *testing.T) {
...
+
+	first := startGateway()
+	second := startGateway()
+	requireEventually(t, 5*time.Second, func() bool {
+		candidates, _ := routes.Candidates(ctx, tunnelID)
+		return len(candidates) == 2 && first.ActiveSessions() == 1 && second.ActiveSessions() == 1 &&
+			len(routes.connections(tunnelID)) == 2
+	})
+
+	require.Equal(t, 1, first.RevokeTunnel(ctx, tunnelID))
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/e2e/tunnel_test.go:363]: Fixed in c943cd73d9: the test now stops the second gateway's agent and drains that gateway before asserting store emptiness — deterministic, no race window.

File: tunnel/gateway/reconciler.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+						slog.String("tunnel_id", tunnelID), slog.Any("error", err))
+				}
+				cancel()
+				return result
+			}
+			cancel()
+			if !active {
+				r.reg.kill(tunnelID)
+				result.cleanupRan = true
+				result.cleaned = r.cleanupTunnel(ctx, tunnelID)
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/reconciler.go]: Fixed in c943cd73d9: drain-path cleanup gives each store mutation its own bounded context (the normal disconnect path keeps the single shared budget per the earlier review request; comments document the asymmetry).

File: tunnel/gateway/gateway.go [hunk 16]
```diff
@@ -437,18 +404,15 @@ func (g *Gateway) newSessionProxy(tunnelID string, session *yamux.Session) http.
...
 	}
-	opCtx, cancel := routeOperationContext(ctx)
-	defer cancel()
-	_ = g.routes.Delete(opCtx, tunnelID)
-	if store, ok := g.routes.(route.ConnectionSnapshotStore); ok {
-		_ = store.DeleteConnections(opCtx, tunnelID)
-	}
-	return g.reg.kill(tunnelID)
+	killed := g.reg.kill(tunnelID)
+	g.reconciler.requestRevoke(ctx, tunnelID)
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/gateway.go:414]: Fixed in c67226b739: the pending global revoke is registered with the reconciler regardless of caller cancellation (tick retry guarantees eventual cleanup); only the synchronous wait honors the caller's ctx. Test covers the canceled-context path.

File: tunnel/gateway/gateway_test.go [hunk 17]
```diff
@@ -36,6 +36,12 @@ func newForwardTestGateway(t *testing.T, cfg Config) *Gateway {
 	t.Helper()
 	gw, err := New(cfg, NewStaticKeyStore(map[string]string{}), route.NewRouteTable(), slog.New(slog.NewTextHandler(io.Discard, nil)))
 	require.NoError(t, err)
+	t.Cleanup(func() {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+		gw.Drain(ctx)
+		gw.CloseSessions(ctx)
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/gateway_test.go]: Fixed in c67226b739: the gateway test cleanup helper now gives CloseSessions its own fresh context, matching the e2e helper.

File: tunnel/gateway/reconciler.go [hunk 18]
```diff
@@ -0,0 +1,430 @@
...
+			return
+		default:
+		}
+
+		select {
+		case req := <-r.drain:
+			r.finishDrain(req, everOwned)
+			return
+		case <-r.nudges:
+			r.reconcile(r.lifetimeCtx, r.takeDirty(), false, everOwned, revokes)
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/gateway/reconciler.go:182]: Production revocation is DB-side (RevokeTunnel currently has no non-test callers), and detection latency for that is inherently one refresh interval regardless of dirty passes: the ticker pass reconciles liveTunnelIDs ∪ everOwned, runs IsActive per key hash, kills inactive sessions, and cleanupTunnel then explicitly deletes the route key (ZREM/HDEL, not a TTL lapse) — so a dirty-pass republish between ticks cannot outlive the next tick. Max staleness is refreshInterval (15s), identical to main's per-session refresher cadence (routeTTL/2). The RevokeTunnel path is already suppressed on every pass: the revokes map is checked before job dispatch independent of validateActive, and reactivation requires a validated active session on a ticker pass. Running IsActive on the dirty pass would put a DB roundtrip on the connect/disconnect hot path (the exact per-session overhead this PR removes) without shrinking that window.

File: tunnel/e2e/tunnel_test.go [hunk 19]
```diff
@@ -305,6 +309,67 @@ func TestTunnelRevoke(t *testing.T) {
...
+	stopSecond()
+	requireEventually(t, 5*time.Second, func() bool { return second.ActiveSessions() == 0 })
+	second.Drain(ctx)
+	candidates, err := routes.Candidates(ctx, tunnelID)
+	require.NoError(t, err)
+	require.Empty(t, candidates)
```
[the remark this answers is not in the document]
Comment (ThomasRooney) [reply tunnel/e2e/tunnel_test.go:369]: requestRevoke is synchronous, not fire-and-forget: the run loop's revoke case executes the global Delete + DeleteConnections before closing the request's done channel, and with the live test context the only other select arms (reconciler exit, ctx cancellation) never fire — so line 363 returns only once the store is globally empty for the tunnel. first cannot republish afterwards: reg.kill is synchronous, the key is revoked from the shared store (agent-1 reconnects fail auth), the revokes map suppresses publish on its disconnect nudge, and its ticker has no live session or everOwned entry left. Any republish by second between lines 363 and 366 goes through its serialized reconciler, which adds the tunnel to everOwned on every publish; second.Drain's final sweep covers everOwned ∪ snapshot and blocks until the reconciler goroutine exits, so no second-side write survives it. The assertions are therefore deterministic without requireEventually.
